retrieval/discovery/gce.go

```diff
@@ -0,0 +1,198 @@
+	gceLabelNetwork      = gceLabel + "network"
+	gceLabelSubnetwork   = gceLabel + "subnetwork"
+	gceLabelPublicIP     = gceLabel + "public_ip"
+	gceLabelPrivateIP    = gceLabel + "private_ip"
+	gceLabelInstanceName = gceLabel + "instance_name"
+	gceLabelTags         = gceLabel + "tags"
+
+	// Constants for instrumentation.
+	namespace = "prometheus"
+)
+
+var (
+	gceSDScrapesCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "gce_sd_scrapes_total",
+			Help:      "The number of GCE-SD scrapes.",
+		})
+	gceSDScrapeFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "gce_sd_scrape_failures_total",
+			Help:      "The number of GCE-SD scrape failures.",
+		})
+	gceSDScrapeDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: namespace,
+			Name:      "gce_sd_scrape_duration",
+			Help:      "The duration of a GCE-SD scrape in seconds.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(gceSDScrapesCount)
+	prometheus.MustRegister(gceSDScrapeFailuresCount)
+	prometheus.MustRegister(gceSDScrapeDuration)
+}
+
+// GCEDiscovery periodically performs GCE-SD requests. It implements
+// the TargetProvider interface.
+type GCEDiscovery struct {
+	project      string
+	zone         string
+	filter       string
+	client       *http.Client
+	svc          *compute.Service
+	isvc         *compute.InstancesService
+	interval     time.Duration
+	port         int
+	tagSeparator string
+}
+
+// NewGCEDiscovery returns a new GCEDiscovery which periodically refreshes its targets.
+func NewGCEDiscovery(conf *config.GCESDConfig) (*GCEDiscovery, error) {
+	gd := &GCEDiscovery{
+		project:      conf.Project,
+		zone:         conf.Zone,
+		filter:       conf.Filter,
+		interval:     time.Duration(conf.RefreshInterval),
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+	var err error
+	gd.client, err = google.DefaultClient(oauth2.NoContext, compute.ComputeReadonlyScope)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up communication with GCE service: %s", err)
+	}
+	gd.svc, err = compute.New(gd.client)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up communication with GCE service: %s", err)
+	}
+	gd.isvc = compute.NewInstancesService(gd.svc)
+	return gd, nil
+}
+
+// Run implements the TargetProvider interface.
+func (gd *GCEDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	// Get an initial set right away.
+	t0 := time.Now()
+	tg, err := gd.refresh()
+	gceSDScrapeDuration.Observe(time.Since(t0).Seconds())
+	gceSDScrapesCount.Inc()
+	if err != nil {
+		gceSDScrapeFailuresCount.Inc()
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	ticker := time.NewTicker(gd.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			t0 := time.Now()
+			tg, err := gd.refresh()
+			gceSDScrapeDuration.Observe(time.Since(t0).Seconds())
```

Comment:
You could remove the duplicated code by moving the observation into `refresh()` (you'll need named return parameters):

```go
func (gd *GCEDiscovery) refresh() (tg *config.TargetGroup, err error) {
	t0 := time.Now()
	defer func() {
		gceSDScrapeDuration.Observe(time.Since(t0).Seconds())
		gceSDScrapesCount.Inc()
		if err != nil {
			gceSDScrapeFailuresCount.Inc()
		}
	}()

	tg.Source = fmt.Sprintf("GCE_%s_%s", gd.project, gd.zone)
	// ...
}
```